### PR TITLE
[JENKINS-55804] - Attempt to solve issue

### PR DESCRIPTION
**Pull request by deepanshnagaria**
my attempt to fix the issue(55804,Performance: Reduce memory footprint and execution time for RoleMap#newMatchingRoleMap()	)by removing the intermediate collections in the method newMatchingRoleMap(), which were created on call of getMatchingRoles() fuction.

### CODE REVIEW

File: src/main/java/com/michelin/cio/hudson/plugins/rolestrategy/RoleMap.java [hunk 1]
```diff
@@ -60,6 +44,14 @@
 import org.kohsuke.stapler.DataBoundConstructor;
 import org.springframework.dao.DataAccessException;
 
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+import java.util.*;
```
Comment (oleg-nenashev) [inline src/main/java/com/michelin/cio/hudson/plugins/rolestrategy/RoleMap.java]: Better to avoid wildcard imports. Sometimes it causes ambiguuty conflicts

**Comment by deepanshnagaria** [comment]
@oleg-nenashev thanks for the response. I was myself going to ask if that method is to be kept or not..... , removed the wildcard, any other changes you would like?

File: src/main/java/com/michelin/cio/hudson/plugins/rolestrategy/RoleMap.java [hunk 2]
```diff
@@ -336,12 +336,18 @@ public void removeRole(Role role){
...
    */
+
   public RoleMap newMatchingRoleMap(String namePattern) {
-    Set<Role> roles = getMatchingRoles(namePattern);
     SortedMap<Role, Set<String>> roleMap = new TreeMap<>();
-    for (Role role : roles) {
-      roleMap.put(role, this.grantedRoles.get(role));
-    }
+    new RoleWalker()
+    {
```
Comment (oleg-nenashev) [inline src/main/java/com/michelin/cio/hudson/plugins/rolestrategy/RoleMap.java]: NIT: codestyle. In Java opening bracket usually goes to the initiating closure, e.g. `new RoleWalker() {`

Comment (deepanshnagaria) [reply src/main/java/com/michelin/cio/hudson/plugins/rolestrategy/RoleMap.java]: Oh i didn't notice that, thanks for pointing. Did the changes, if any further changes are required do point out. Thanks. @oleg-nenashev

File: src/main/java/com/michelin/cio/hudson/plugins/rolestrategy/RoleMap.java [hunk 3]
```diff
@@ -336,12 +336,17 @@ public void removeRole(Role role){
...
   public RoleMap newMatchingRoleMap(String namePattern) {
-    Set<Role> roles = getMatchingRoles(namePattern);
     SortedMap<Role, Set<String>> roleMap = new TreeMap<>();
-    for (Role role : roles) {
-      roleMap.put(role, this.grantedRoles.get(role));
-    }
+    new RoleWalker() {
+      public void perform(Role current) {
+        Matcher m = current.getPattern().matcher(namePattern);
+          if (m.matches()) {
```
Comment (oleg-nenashev) [inline src/main/java/com/michelin/cio/hudson/plugins/rolestrategy/RoleMap.java]: Minor thing: indenting is broken here

Comment (deepanshnagaria) [reply src/main/java/com/michelin/cio/hudson/plugins/rolestrategy/RoleMap.java]: Sorry for those repeated formatting issues. I have changed that, will try to overcome this next time on wards.

**Comment by oleg-nenashev** [comment]
Thanks @deepanshnagaria !

**Comment by oleg-nenashev** [comment]
Uh, I finally didn't squash it :(
Anyway, it is not a big traffic in the plugin nowadays

**Comment by deepanshnagaria** [comment]
@oleg-nenashev  i wanted to ask if the implementation i did is what also solves https://issues.jenkins-ci.org/browse/JENKINS-55807 ? Actually i didn't understand the issue clearly ,but feel that it says something similar..

**Pull request merged**